presto-postgresql/src/test/java/io/prestosql/plugin/postgresql/TestPostgreSqlTypeMapping.java

```diff
@@ -361,6 +375,168 @@ public void testDecimalExceedingPrecisionMaxConvertedToVarchar()
+            }
+            for (int scale : asList(29, 38)) {
+                mappingScale = calcDecimalScaleMapping(typePrecision, typeScale, scale);
+                assertQuery(sessionWithDecimalMappingAllowOverflow(UNNECESSARY, scale), selectSchema, expectedSchema.apply(mappingScale));
+                assertQuery(sessionWithDecimalMappingAllowOverflow(UNNECESSARY, scale), selectRows.apply("exact"), expectedValues.apply("exact"));
+                assertQueryFails(sessionWithDecimalMappingAllowOverflow(UNNECESSARY, scale), selectSupportedRows, "Decimal overflow");
+                assertQuery(sessionWithDecimalMappingAllowOverflow(HALF_UP, scale), selectSchema, expectedSchema.apply(mappingScale));
+                assertQueryFails(sessionWithDecimalMappingAllowOverflow(HALF_UP, scale), selectSupportedRows, "Decimal overflow");
+            }
+
+            assertQuery(sessionWithDecimalMappingStrict(CONVERT_TO_VARCHAR), selectSchema, "VALUES ('label','varchar(12)'), ('d_col', 'varchar')");
+            assertQuery(sessionWithDecimalMappingStrict(CONVERT_TO_VARCHAR), selectRows.apply("%"),
+                    "VALUES " + testData.stream()
+                            .map(s -> s.split(",")[1].trim())
+                            .map(s -> format("('%s')", StringUtils.rightPad(s, s.split("[.]")[0].length() + typeScale + 1, "0")))
+                            .collect(joining(",")));
+
+            assertQuery(sessionWithDecimalMappingStrict(IGNORE), selectSchema, "VALUES ('label','varchar(12)')");
+        }
+    }
+
+    @Test
+    public void testDecimalUnspecifiedPrecisionWithSupportedValues()
+    {
+        JdbcSqlExecutor jdbcSqlExecutor = new JdbcSqlExecutor(postgreSqlServer.getJdbcUrl());
+        List<String> testData = asList("1.12", "123456.789", "-1.12", "-123456.789");
+        try (TestTable testTable = new TestTable(
+                jdbcSqlExecutor,
+                "tpch.test_var_decimal",
+                "(d_col decimal,remark varchar)",
+                testData)) {
+            String selectSchema = format("SELECT column_name, data_type FROM information_schema.columns WHERE table_schema = 'tpch' AND table_name = '%s'", testTable.getName().split("[.]")[1]);
+            String selectRows = format("SELECT d_col FROM %s", testTable.getName());
+            Function<Integer, String> expectedSchema = (scale) -> format("VALUES ('d_col','decimal(%d,%d)'),('remark','varchar')", Decimals.MAX_PRECISION, scale);
+            String expectedExactValues = "VALUES " + testData.stream().collect(joining(","));
+            Function<List<Object>, String> expectedScaledValues = (rounding) -> "VALUES " + testData.stream()
+                    .map(s -> new BigDecimal(s).setScale((Integer) rounding.get(0), (RoundingMode) rounding.get(1)).toString())
+                    .collect(joining(","));
+            int largestScale = testData.stream().map(e -> e.length() - e.indexOf(".") - 1).max(Integer::compareTo).orElse(0);
```

Comment:
```
            int largestScale = testData.stream()
                            .map(e -> e.length() - e.indexOf(".") - 1)
                            .max(Integer::compareTo)
                            .orElse(0);
````

That way it is much better to read it. It is like having separate statement in separate line. Use it for most of the usages of `.stream()`.